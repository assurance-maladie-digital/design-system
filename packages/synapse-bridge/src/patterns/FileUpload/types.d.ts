import type { ErrorCodes } from './errorCodes';

export interface ErrorEvent {
  file: File | DataTransferItemList
  code: ErrorCodes
}

export interface ErrorCodesType {
  [key: string]: string
}

export interface HTMLInputEvent extends Event {
  target: HTMLInputElement & EventTarget
}
