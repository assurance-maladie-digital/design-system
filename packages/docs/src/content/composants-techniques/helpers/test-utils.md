---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
