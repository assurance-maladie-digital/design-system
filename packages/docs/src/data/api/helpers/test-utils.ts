import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
