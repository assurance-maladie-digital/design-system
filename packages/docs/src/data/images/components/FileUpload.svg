<svg width="541" height="357" viewBox="0 0 541 357" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect width="541" height="357" fill="#E7ECF5"/>
<rect x="63" y="88" width="415" height="181" fill="white"/>
<path d="M266.794 193.168C266.243 193.168 265.725 193.028 265.24 192.748C264.764 192.468 264.377 192.062 264.078 191.53C263.789 190.998 263.644 190.359 263.644 189.612C263.644 188.847 263.789 188.198 264.078 187.666C264.377 187.134 264.764 186.728 265.24 186.448C265.725 186.168 266.243 186.028 266.794 186.028C267.354 186.028 267.872 186.168 268.348 186.448C268.824 186.728 269.207 187.134 269.496 187.666C269.795 188.198 269.944 188.847 269.944 189.612C269.944 190.359 269.795 190.998 269.496 191.53C269.207 192.062 268.824 192.468 268.348 192.748C267.872 193.028 267.354 193.168 266.794 193.168ZM266.794 192.216C267.382 192.216 267.853 191.978 268.208 191.502C268.572 191.017 268.754 190.387 268.754 189.612C268.754 188.828 268.572 188.193 268.208 187.708C267.853 187.223 267.382 186.98 266.794 186.98C266.215 186.98 265.744 187.223 265.38 187.708C265.016 188.193 264.834 188.828 264.834 189.612C264.834 190.387 265.016 191.017 265.38 191.502C265.744 191.978 266.215 192.216 266.794 192.216ZM273.724 193.168C273.005 193.168 272.478 192.944 272.142 192.496C271.806 192.039 271.638 191.376 271.638 190.508V186.196H272.8V190.354C272.8 190.989 272.898 191.451 273.094 191.74C273.299 192.029 273.626 192.174 274.074 192.174C274.429 192.174 274.741 192.085 275.012 191.908C275.292 191.721 275.591 191.427 275.908 191.026V186.196H277.056V193H276.104L276.006 191.936H275.964C275.647 192.309 275.311 192.608 274.956 192.832C274.601 193.056 274.191 193.168 273.724 193.168Z" fill="black" fill-opacity="0.45"/>
<g filter="url(#filter0_ddd_213_106951)">
<g clip-path="url(#clip0_213_106951)">
<rect x="229.5" y="211" width="80" height="36.0003" rx="4" fill="#0C419A"/>
</g>
</g>
<rect opacity="0.5" x="177" y="153" width="191" height="19" rx="9.5" fill="#CED9EB"/>
<g clip-path="url(#clip1_213_106951)">
<path d="M282.25 116.733C281.117 110.983 276.067 106.667 270 106.667C265.183 106.667 261 109.4 258.917 113.4C253.9 113.933 250 118.183 250 123.333C250 128.85 254.483 133.333 260 133.333H281.667C286.267 133.333 290 129.6 290 125C290 120.6 286.583 117.033 282.25 116.733ZM273.333 121.667V128.333H266.667V121.667H261.667L270 113.333L278.333 121.667H273.333Z" fill="#0C419A"/>
</g>
<rect opacity="0.5" x="241" y="219" width="57" height="19" rx="9.5" fill="#CED9EB"/>
<defs>
<filter id="filter0_ddd_213_106951" x="224.5" y="207" width="90" height="46.0002" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB">
<feFlood flood-opacity="0" result="BackgroundImageFix"/>
<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/>
<feOffset dy="1"/>
<feGaussianBlur stdDeviation="2.5"/>
<feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.12 0"/>
<feBlend mode="normal" in2="BackgroundImageFix" result="effect1_dropShadow_213_106951"/>
<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/>
<feOffset dy="2"/>
<feGaussianBlur stdDeviation="1"/>
<feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.14 0"/>
<feBlend mode="normal" in2="effect1_dropShadow_213_106951" result="effect2_dropShadow_213_106951"/>
<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/>
<feMorphology radius="2" operator="erode" in="SourceAlpha" result="effect3_dropShadow_213_106951"/>
<feOffset dy="3"/>
<feGaussianBlur stdDeviation="0.5"/>
<feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.2 0"/>
<feBlend mode="normal" in2="effect2_dropShadow_213_106951" result="effect3_dropShadow_213_106951"/>
<feBlend mode="normal" in="SourceGraphic" in2="effect3_dropShadow_213_106951" result="shape"/>
</filter>
<clipPath id="clip0_213_106951">
<rect x="229.5" y="211" width="80" height="36.0003" rx="4" fill="white"/>
</clipPath>
<clipPath id="clip1_213_106951">
<rect width="40" height="40" fill="white" transform="translate(250 100)"/>
</clipPath>
</defs>
</svg>
