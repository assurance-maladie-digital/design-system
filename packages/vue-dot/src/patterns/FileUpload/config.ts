export const config = {
	icon: {
		size: 40,
		color: 'primary'
	}
};
