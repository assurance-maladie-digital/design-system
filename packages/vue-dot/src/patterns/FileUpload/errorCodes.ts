export enum ErrorCodes {
	MULTIPLE_FILES_SELECTED = 'MULTIPLE_FILES_SELECTED',
	FILE_TOO_LARGE = 'FILE_TOO_LARGE',
	FILE_EXTENSION_NOT_ALLOWED = 'FILE_EXTENSION_NOT_ALLOWED'
}
