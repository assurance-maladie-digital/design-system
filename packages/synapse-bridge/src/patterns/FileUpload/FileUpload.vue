<script lang="ts">
import type {PropType} from "vue";
import {defineComponent} from "vue";

import {config} from "./config";
import {locales} from "./locales";

import FileUploadCore from "./mixins/fileUploadCore";
import {Widthable} from "@/mixins/widthable";

import {customizable} from "@/mixins/customizable";
import {calcHumanFileSize} from "@/functions/calcHumanFileSize";

import {mdiCloudUpload} from "@mdi/js";

import type {IndexedObject} from "@/types";
import type {HTMLInputEvent} from "@/patterns/FileUpload/types";

export default defineComponent({
	mixins: [customizable(config), FileUploadCore, Widthable],
	props: {
		modelValue: {
			// File is not a valid prop type,
			// use null to allow any type & provide custom validation
			type: null as unknown as PropType<File | File[] | null>,
			default: null,
			/** @see https://github.com/vuetifyjs/vuetify/blob/9e4daa81bfe3046fcf7e4a731809c3d554b3df54/packages/vuetify/src/components/VFileInput/VFileInput.tsx#L72C13-L72C13 */
			validator(val): boolean {
				const value = Array.isArray(val) ? val : [val];
				return value.every(
					(v) => v !== null && typeof v === "object"
				);
			},
		},
		multiple: {
			type: Boolean,
			default: false,
		},
		noRipple: {
			type: Boolean,
			default: false,
		},
		disabled: {
			type: Boolean,
			default: false,
		},
	},
	emits: ["update:modelValue"],
	data() {
		return {
			locales,
			uploadIcon: mdiCloudUpload
		};
	},
	computed: {
		colors(): IndexedObject {
			return {
				label: 'text-primary',
				multiple: 'text-black',
				info: 'text-grey'
			};
		},
		maxSizeReadable(): string {
			return calcHumanFileSize(this.fileSizeMax, this.fileSizeUnits);
		},
		extensions(): string {
			return this.allowedExtensions.join(', ').toUpperCase();
		}
	},
	methods: {
		retry(): void {
			this.$refs.vdInputEl.click();
		}
	}
});
</script>

<template>
	<label
		v-ripple="!noRipple"
		:style="widthStyles"
		:class="[
			{
				dragover: dragover,
				'dark-mode': $vuetify.theme.current.dark,
			},
			colors.label,
		]"
		class="vd-file-upload d-block pa-4"
		@dragover.prevent="dragover = true"
		@dragleave="dragover = false"
		@drop.prevent="disabled ? null : dropHandler($event)"
	>
		<input
			ref="vdInputEl"
			type="file"
			:disabled="disabled"
			:multiple="multiple"
			:accept="computedAccept"
			class="vd-file-upload-input"
			@change="e=>inputValueChanged(e as HTMLInputEvent)"
		/>

		<slot name="placeholder">
			<span class="vd-file-upload-placeholder">
				<slot name="icon">
					<VIcon v-bind="options.icon">
						{{ uploadIcon }}
					</VIcon>
				</slot>

				<span :class="colors.multiple" class="mt-1 font-weight-medium">
					<slot name="action-text" :multiple="multiple">
						{{
							multiple
								? "Placez vos fichiers ici"
								: "Placez votre fichier ici"
						}}
					</slot>
				</span>

				<span :class="colors.info" class="mb-2">
					<slot name="or">
						{{ locales.or }}
					</slot>
				</span>

				<span
					class="vd-file-upload-btn bg-primary text-white text-uppercase py-2 px-4 elevation-2"
				>
					<slot name="button-text">
						{{ locales.chooseFile }}
					</slot>
				</span>

				<span
					:class="colors.info"
					class="mt-4 text-body-2 font-weight-regular"
				>
					<slot
						name="info-text"
						:max-size="maxSizeReadable"
						:extensions="extensions"
					>
						{{
							locales.infoText(
								maxSizeReadable,
								extensions,
								allowedExtensions.length
							)
						}}
					</slot>
				</span>
			</span>
		</slot>
	</label>
</template>

<style lang="scss" scoped>
@import "@cnamts/design-tokens/dist/tokens";
.vd-file-upload {
	cursor: pointer;
	position: relative;
	border: 1px dashed;
	transition: background 0.25s;
	&:hover,
	&:focus-within,
	&.dragover {
		background: $vd-parme-lighten-90;
	}
	&.dark-mode {
		&:hover,
		&:focus-within,
		&.dragover {
			background: #303030;
		}
	}
}
.vd-file-upload-placeholder {
	display: flex;
	align-items: center;
	flex-direction: column;
}
.vd-file-upload-input {
	position: absolute;
	width: 1px;
	height: 1px;
	white-space: nowrap;
	overflow: hidden;
	clip: rect(1px, 1px, 1px, 1px);
}
.vd-file-upload-btn {
	border-radius: 4px;
	transition: background 0.25s;
}
input[disabled] ~ .vd-file-upload-placeholder {
	opacity: 0.2;
}
</style>
