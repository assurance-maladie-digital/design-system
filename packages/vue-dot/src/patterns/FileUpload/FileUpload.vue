<template>
	<label
		v-ripple="!noRipple"
		:style="widthStyles"
		:class="[
			{
				'dragover': dragover,
				'dark-mode': $vuetify.theme.dark
			},
			colors.label
		]"
		class="vd-file-upload d-block pa-4"
		@dragover.prevent="dragover = true"
		@dragleave="dragover = false"
		@drop.prevent="dropHandler"
	>
		<input
			ref="vdInputEl"
			type="file"
			:disabled="disabled"
			:multiple="multiple"
			:accept="computedAccept"
			class="vd-file-upload-input"
			@change="inputValueChanged"
		>

		<slot name="placeholder">
			<span class="vd-file-upload-placeholder">
				<slot name="icon">
					<VIcon v-bind="options.icon">
						{{ uploadIcon }}
					</VIcon>
				</slot>

				<span
					:class="colors.multiple"
					class="mt-1 font-weight-medium"
				>
					<slot
						name="action-text"
						:multiple="multiple"
					>
						{{ multiple ? 'Placez vos fichiers ici' : 'Placez votre fichier ici' }}
					</slot>
				</span>

				<span
					:class="colors.info"
					class="mb-2"
				>
					<slot name="or">
						{{ locales.or }}
					</slot>
				</span>

				<span class="vd-file-upload-btn primary white--text text-uppercase py-2 px-4 elevation-2">
					<slot name="button-text">
						{{ locales.chooseFile }}
					</slot>
				</span>

				<span
					:class="colors.info"
					class="mt-4 text-body-2 font-weight-regular"
				>
					<slot
						name="info-text"
						:max-size="maxSizeReadable"
						:extensions="extensions"
					>
						{{ locales.infoText(maxSizeReadable, extensions, allowedExtensions.length) }}
					</slot>
				</span>
			</span>
		</slot>
	</label>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { config } from './config';
	import { locales } from './locales';

	import { FileUploadCore } from './mixins/fileUploadCore';
	import { Widthable } from '../../mixins/widthable';

	import { customizable } from '../../mixins/customizable';

	import { calcHumanFileSize } from '../../functions/calcHumanFileSize';

	import { mdiCloudUpload } from '@mdi/js';

	import { IndexedObject } from '../../types';

	const Props = Vue.extend({
		props: {
			value: {
				// File is not a valid prop type,
				// use null to allow any type & provide custom validation
				type: null as unknown as PropType<File | File[] | null>,
				default: null,
				/** @see https://github.com/vuetifyjs/vuetify/blob/master/packages/vuetify/src/components/VFileInput/VFileInput.ts#L71 */
				validator(val): boolean {
					if (val === null) {
						return true;
					}

					const value = Array.isArray(val) ? val : [val];
					const isValid = value.every((v) => v !== null && typeof v === 'object');

					return isValid;
				}
			},
			multiple: {
				type: Boolean,
				default: false
			},
			noRipple: {
				type: Boolean,
				default: false
			},
			disabled: {
				type: Boolean,
				default: false
			}
		}
	});

	const MixinsDeclaration = mixins(Props, customizable(config), FileUploadCore, Widthable);

	@Component({
		model: {
			prop: 'value',
			event: 'change'
		}
	})
	export default class FileUpload extends MixinsDeclaration {
		locales = locales;

		uploadIcon = mdiCloudUpload;

		get colors(): IndexedObject {
			return {
				label: 'primary--text',
				multiple: 'black--text',
				info: 'grey--text'
			};
		}

		get maxSizeReadable(): string {
			return calcHumanFileSize(this.fileSizeMax, this.fileSizeUnits);
		}

		get extensions(): string {
			return this.allowedExtensions.join(', ').toUpperCase();
		}

		public retry(): void {
			this.$refs.vdInputEl.click();
		}
	}
</script>

<style lang="scss" scoped>
	@import '@cnamts/design-tokens/dist/tokens';

	.vd-file-upload {
		cursor: pointer;
		position: relative;
		border: 1px dashed;
		background: #fff;
		transition: background .25s;

		&:hover,
		&:focus-within,
		&.dragover {
			background: $vd-parme-lighten-90;
		}

		&.dark-mode {
			&:hover,
			&:focus-within,
			&.dragover {
				background: #303030;
			}
		}
	}

	.vd-file-upload-placeholder {
		display: flex;
		align-items: center;
		flex-direction: column;
	}

	.vd-file-upload-input {
		position: absolute;
		width: 1px;
		height: 1px;
		white-space: nowrap;
		overflow: hidden;
		clip: rect(1px, 1px, 1px, 1px);
	}

	.vd-file-upload-btn {
		border-radius: 4px;
		transition: background .25s;
	}
</style>
