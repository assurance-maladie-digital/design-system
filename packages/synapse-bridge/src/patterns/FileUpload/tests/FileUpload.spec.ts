import { describe, it, expect, vi, afterEach } from "vitest";
import { shallowMount } from "@vue/test-utils";

import FileUpload from "../";
import { vuetify } from "@tests/unit/setup";

const file = new File([""], "filename", { type: "text/html" });

describe("FileUpload", () => {
	const consoleMock = vi.spyOn(console, 'warn').mockImplementation(() => undefined);

	afterEach(() => {
		consoleMock.mockReset();
	});

	it("renders correctly", () => {
		const wrapper = shallowMount(FileUpload, {
			global: {
				plugins: [vuetify],
			},
		});

		expect(wrapper).toMatchSnapshot();
	});

	it("renders correctly in multiple mode", () => {
		const wrapper = shallowMount(FileUpload, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				multiple: true,
				modelValue: [file, file],
			},
		});

		expect(wrapper).toMatchSnapshot();
	});

	it("renders correctly with only one extension allowed", () => {
		const wrapper = shallowMount(FileUpload, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				allowedExtensions: ["pdf"],
				modelValue: file,
			},
		});

		expect(wrapper).toMatchSnapshot();
	});

	it("do not accept invalid props", () => {
		shallowMount(FileUpload, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				// @ts-expect-error
				modelValue: [null],
			},
		});

		const mockArg: null | string = consoleMock.mock.calls?.[0]?.[0];
		expect(mockArg).toContain("[Vue warn]: Invalid prop: custom validator check failed for prop \"modelValue\"");
	});

	it('accepts drag and drop', () => {
		const wrapper = shallowMount(FileUpload, {
			global: {
				plugins: [vuetify],
			},
		});

		const mockDropHandler= vi.fn();
		const dropHandler = wrapper.vm.dropHandler;
		wrapper.vm.dropHandler = mockDropHandler;

		wrapper.find('input').trigger('drop');
		expect(mockDropHandler).toHaveBeenCalled();

		wrapper.vm.dropHandler = dropHandler;
	});

	it('change the style when dragging', async() => {
		const wrapper = shallowMount(FileUpload, {
			global: {
				plugins: [vuetify],
			},
		});

		const label = wrapper.find('label');
		const labelClasses = label.classes();

		await wrapper.find('input').trigger('dragover');
		expect(label.classes()).not.toEqual(labelClasses);

		await wrapper.find('input').trigger('dragleave');
		expect(label.classes()).toEqual(labelClasses);
	});

	it('call the mixin when the input value changes', async() => {
		const wrapper = shallowMount(FileUpload, {
			global: {
				plugins: [vuetify],
			},
		});

		const mockInputValueChanged = vi.fn();
		const inputValueChanged = wrapper.vm.inputValueChanged;
		wrapper.vm.inputValueChanged = mockInputValueChanged;

		await wrapper.find('input').trigger('change');
		expect(mockInputValueChanged).toHaveBeenCalled();

		wrapper.vm.inputValueChanged = inputValueChanged;
	});

	it('retries the upload of a file', () => {
		const wrapper = shallowMount(FileUpload, {
			global: {
				plugins: [vuetify],
			},
		});

		const mockRetry = vi.fn();
		wrapper.vm.retry = mockRetry;

		wrapper.vm.retry();

		expect(mockRetry).toHaveBeenCalled();
	});

	it('clicks the input', () => {
		const wrapper = shallowMount(FileUpload, {
			global: {
				plugins: [vuetify],
			},
		});

		const mockClick = vi.fn();
		wrapper.vm.$refs.vdInputEl.click = mockClick;

		wrapper.vm.retry();

		expect(mockClick).toHaveBeenCalled();
	});
});

