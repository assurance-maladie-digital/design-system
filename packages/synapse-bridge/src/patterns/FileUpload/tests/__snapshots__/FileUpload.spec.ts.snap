// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileUpload > renders correctly 1`] = `<label data-v-b5ec5665="" style="width: 100%;" class="text-primary vd-file-upload d-block pa-4"><input data-v-b5ec5665="" type="file" accept=".pdf,.jpg,.jpeg,.png" class="vd-file-upload-input"><span data-v-b5ec5665="" class="vd-file-upload-placeholder"><v-icon-stub data-v-b5ec5665="" color="primary" disabled="false" start="false" end="false" size="40" tag="i"></v-icon-stub><span data-v-b5ec5665="" class="text-black mt-1 font-weight-medium">Placez votre fichier ici</span><span data-v-b5ec5665="" class="text-grey mb-2">Ou</span><span data-v-b5ec5665="" class="vd-file-upload-btn bg-primary text-white text-uppercase py-2 px-4 elevation-2">Choisir un fichier</span><span data-v-b5ec5665="" class="text-grey mt-4 text-body-2 font-weight-regular">Taille max. : 10 Mo. Formats acceptés : PDF, JPG, JPEG, PNG</span></span></label>`;

exports[`FileUpload > renders correctly in multiple mode 1`] = `<label data-v-b5ec5665="" style="width: 100%;" class="text-primary vd-file-upload d-block pa-4"><input data-v-b5ec5665="" type="file" multiple="" accept=".pdf,.jpg,.jpeg,.png" class="vd-file-upload-input"><span data-v-b5ec5665="" class="vd-file-upload-placeholder"><v-icon-stub data-v-b5ec5665="" color="primary" disabled="false" start="false" end="false" size="40" tag="i"></v-icon-stub><span data-v-b5ec5665="" class="text-black mt-1 font-weight-medium">Placez vos fichiers ici</span><span data-v-b5ec5665="" class="text-grey mb-2">Ou</span><span data-v-b5ec5665="" class="vd-file-upload-btn bg-primary text-white text-uppercase py-2 px-4 elevation-2">Choisir un fichier</span><span data-v-b5ec5665="" class="text-grey mt-4 text-body-2 font-weight-regular">Taille max. : 10 Mo. Formats acceptés : PDF, JPG, JPEG, PNG</span></span></label>`;

exports[`FileUpload > renders correctly with only one extension allowed 1`] = `<label data-v-b5ec5665="" style="width: 100%;" class="text-primary vd-file-upload d-block pa-4"><input data-v-b5ec5665="" type="file" accept=".pdf" class="vd-file-upload-input"><span data-v-b5ec5665="" class="vd-file-upload-placeholder"><v-icon-stub data-v-b5ec5665="" color="primary" disabled="false" start="false" end="false" size="40" tag="i"></v-icon-stub><span data-v-b5ec5665="" class="text-black mt-1 font-weight-medium">Placez votre fichier ici</span><span data-v-b5ec5665="" class="text-grey mb-2">Ou</span><span data-v-b5ec5665="" class="vd-file-upload-btn bg-primary text-white text-uppercase py-2 px-4 elevation-2">Choisir un fichier</span><span data-v-b5ec5665="" class="text-grey mt-4 text-body-2 font-weight-regular">Taille max. : 10 Mo. Format accepté : PDF</span></span></label>`;
